mpi4jax/collective_ops/sendrecv.py

```diff
@@ -0,0 +1,163 @@
+from jax.lax import create_token
+from jax.core import Primitive
+from jax.lib import xla_client
+from jax.interpreters import xla
+
+from ..utils import (
+    to_mpi_ptr,
+    _unpack_builder,
+    _ops,
+    _constant_s32_scalar,
+    _constant_u64_scalar,
+    dtype_ptr,
+)
+
+from ..warn import warn_missing_omnistaging
+
+# The Jax primitive
+mpi_sendrecv_p = Primitive("sendrecv_mpi")  # Create the primitive
+
+
+# This function applies the primitive to an AST
+def Sendrecv(
+    sendbuf,
+    recvbuf,
+    source,
+    dest,
+    sendtag=0,
+    recvtag=_MPI.ANY_TAG,
+    comm=_MPI.COMM_WORLD,
+    status=None,
+    token=None,
+):
+    if token is None:
+        token = create_token(sendbuf)
+
+    return mpi_sendrecv_p.bind(
+        sendbuf,
+        recvbuf,
+        token,
+        source=source,
+        dest=dest,
+        sendtag=sendtag,
+        recvtag=recvtag,
+        comm=comm,
+        status=status,
+    )
+
+
+#  this function executes the primitive, when not under any transformation
+def mpi_sendrecv_impl(
+    sendbuf, recvbuf, token, source, dest, sendtag, recvtag, comm, status
+):
+    # TODO: make this support gpus (use cupy?)
+    inarr = _np.asarray(sendbuf)
+    outarr = _np.empty_like(recvbuf)
+
+    comm.Sendrecv(
+        sendbuf=inarr,
+        dest=dest,
+        sendtag=sendtag,
+        recvbuf=outarr,
+        source=source,
+        recvtag=recvtag,
+        status=status,
+    )
+
+    res = jnp.array(outarr, dtype=recvbuf.dtype)
+
+    # put the result on the correct device if needed
+    if not (res.device_buffer.device() == recvbuf.device_buffer.device()):
+        res = device_put(res, device=recvbuf.device_buffer.device())
+
+    return res, token
+
+
+#  This function compiles the operation
+def mpi_sendrecv_xla_encode(
+    c, sendbuf, recvbuf, token, source, dest, sendtag, recvtag, comm, status
+):
+    from ..cython.mpi_xla_bridge import MPI_STATUS_IGNORE_ADDR
+
+    warn_missing_omnistaging()
+
+    c = _unpack_builder(c)
+
+    recv_shape = c.GetShape(recvbuf)
+    recv_dtype = recv_shape.element_type()
+    recv_dims = recv_shape.dimensions()
+
+    # compute total number of elements in array
+    recv_nitems = recv_dims[0]
+    for el in recv_dims[1:]:
+        recv_nitems *= el
+
+    _recv_nitems = _constant_s32_scalar(c, recv_nitems)
+    _recv_dtype_ptr = dtype_ptr(recv_dtype)
+
+    send_shape = c.GetShape(sendbuf)
+    send_dtype = send_shape.element_type()
+    send_dims = send_shape.dimensions()
+
+    # compute total number of elements in array
+    send_nitems = send_dims[0]
+    for el in send_dims[1:]:
+        send_nitems *= el
+
+    _send_nitems = _constant_s32_scalar(c, send_nitems)
+    _send_dtype_ptr = dtype_ptr(send_dtype)
+
+    sh = xla_client.Shape.tuple_shape(
+        [
+            xla_client.Shape.array_shape(recv_dtype, recv_dims),
+            xla_client.Shape.token_shape(),
+        ]
+    )
+
+    if status is None:
+        _status = MPI_STATUS_IGNORE_ADDR
```

Comment:
what is this exactly? why do you want to set it externally?

Reply:
If you do not want to use a status with `MPI_Recv` you can pass in `MPI_STATUS_IGNORE`. Unfortunately, `mpi4py` does not wrap `MPI_STATUS_IGNORE`, so it's not available from Python. Instead, I export the address of `MPI_STATUS_IGNORE` in our Cython wrapper.

It doesn't really need to be set externally, but our Cython wrappers expect a memory address to the status, so this seemed like the most convenient solution. Alternatively we could pass something like `nullptr` and check for that in the Cython wrapper (and replace it with `MPI_STATUS_IGNORE`), but that seemed clumsy to me.